File: demo-server/src/ws/session.rs
//! Represents a Websocket session for an Agent.
//! Reminder that the server is the one sending RPC requests to the Agent and receiving replies.
use std::{collections::HashMap, sync::Arc};

use futures_util::{stream::SplitSink, SinkExt};
use poem::web::websocket::{Message, WebSocketStream};
use rpc::{RpcMessage, RpcRequest, RpcResponse};
use tokio::sync::{oneshot, Mutex};
#[derive(Clone)]
pub struct WsSession {
    pub id: uuid::Uuid,
    pub tx: Arc<Mutex<SplitSink<WebSocketStream, Message>>>,
    callbacks: Arc<Mutex<HashMap<uuid::Uuid, oneshot::Sender<RpcResponse>>>>,
}

impl WsSession {
    pub fn new(ws_tx: SplitSink<WebSocketStream, Message>) -> Self {
        let id = uuid::Uuid::new_v4();
        let tx = Arc::new(Mutex::new(ws_tx));
        let callbacks = Arc::new(Mutex::new(HashMap::new()));
        Self { id, tx, callbacks }
    }

    pub async fn handle_message(&self, msg: String) {
        println!("Received message: {}", msg);
        let parsed_msg = serde_json::from_str::<RpcMessage<RpcResponse>>(&msg);
        match parsed_msg {
            Ok(msg) => {
                // check if msg.id is in callbacks
                let mut callbacks = self.callbacks.lock().await;
                if let Some(tx) = callbacks.remove(&msg.id) {
                    println!("Found callback for message: {}", msg.id);
                    let _ = tx.send(msg.payload);
                } else {
                    println!("No callback for message: {}", msg.id);
                }
            }
            Err(e) => {
                println!("Error parsing message: {}", e);
            }
        }
    }

    pub async fn send(&self, text: String) {
        let mut tx = self.tx.lock().await;
        let _ = tx.send(Message::Text(text)).await;
    }

    pub async fn send_rpc(&self, req: RpcRequest) -> RpcResponse {
        let msg = RpcMessage {
            id: uuid::Uuid::new_v4(),
            payload: req,
        };
        let msg_se = serde_json::to_string(&msg).unwrap();
        let (cb_tx, cb_rx) = oneshot::channel::<RpcResponse>();
        let mut callbacks = self.callbacks.lock().await;
        callbacks.insert(msg.id, cb_tx);
        drop(callbacks);
        let mut tx = self.tx.lock().await;
        let _ = tx.send(Message::Text(msg_se)).await;
        drop(tx);
        cb_rx.await.unwrap()
    }
}

File: fastapi-server/app/ws/session.py
import asyncio
import uuid
from typing import Dict, Optional

from app.rpc import Message, RpcRequest, RpcResponse
from fastapi import WebSocket
from pydantic import ValidationError


class WsSession:
    def __init__(self, conn: WebSocket):
        self.id = uuid.uuid4()
        self.conn = conn
        self.callbacks: Dict[uuid.UUID: asyncio.Future] = {}

    async def handle_message(self, msg: str):
        print(f"Received message: {msg}")
        try:
            parsed = Message.model_validate_json(msg)
        except ValidationError as e:
            print(f"Failed parsing to Message: {e}")
            return
        except Exception as e:
            print(f"Unexpected error: {e}")
            return
        cb: Optional[asyncio.Future] = self.callbacks.pop(parsed.id, None)
        if not cb:
            print(f"Received message with no callback: {parsed}")
            return
        cb.set_result(parsed.payload)

    async def send(self, text: str):
        print(f"Sending message: {text}")
        await self.conn.send_text(text)

    async def send_rpc(self, req: RpcRequest) -> RpcResponse:
        msg_id = uuid.uuid4()
        msg = Message(id=msg_id, payload=req)
        fut = asyncio.Future()
        self.callbacks[msg_id] = fut
        await self.send(msg.model_dump_json())
        return await fut



        